### PR TITLE
Replace "assert" by "if raise"

### LINKED ISSUE #277: Replace asserts in the code by if condition then raise
It would be better to replace the code as 

https://github.com/pytorch/ignite/blob/0f1905e87b95779cf6544a5fe29b46519ad9d4e0/ignite/metrics/epoch_metric.py#L33-L34

by 
```python
if y_pred.ndimension()  < 1 or  y_pred.ndimension() > 2:
    raise TypeError("Predictions should be of shape (batch_size, n_classes)"
# etc
```

Usage of assert to change in `EpochMetric`, `EarlyStopping` and `Loss`

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #277  

Description:
Replace "assert" by "if raise" pattern

Check list:
* [x] New tests are added

### CODE REVIEW

File: ignite/handlers/early_stopping.py [hunk 1]
```diff
@@ -32,9 +32,16 @@ def score_function(engine):
...
     def __init__(self, patience, score_function, trainer):
-        assert callable(score_function), "Argument score_function should be a function"
-        assert patience > 0, "Argument patience should be positive"
-        assert isinstance(trainer, Engine), "Argument trainer should be an instance of Engine"
+
+        if not callable(score_function):
+            raise TypeError("Argument score_function should be a function")
+
+        if patience < 1:
+            raise ValueError("Argument patience should be positive")
```
Comment (alykhantejani) [inline ignite/handlers/early_stopping.py]: nit: should be a positive *integer*

File: ignite/metrics/loss.py [hunk 2]
```diff
@@ -38,7 +38,10 @@ def update(self, output):
         else:
             y_pred, y, kwargs = output
         average_loss = self._loss_fn(y_pred, y, **kwargs)
-        assert len(average_loss.shape) == 0, '`loss_fn` did not return the average loss'
+
+        if len(average_loss.shape) != 0:
+            raise ValueError('`loss_fn` did not return the average loss')
```
Comment (alykhantejani) [inline ignite/metrics/loss.py]: can remove the backticks

**Comment by vfdev-5** [comment]
@alykhantejani I fixed what you've asked, please merge this :)
